harvester_e2e_tests/integration/test_upgrade.py

```diff
@@ -1,121 +1,1049 @@
+}
+
+pytest_plugins = [
+    "harvester_e2e_tests.fixtures.api_client",
+]
+
+
+def _lookup_image(api_client, name):
+    code, data = api_client.images.get()
+    if code == 200:
+        for image in data['items']:
+            if image['metadata']['name'] == name:
+                return image
+    return None
+
+
+def _create_image(api_client, url, name, wait_timeout=300):
+    image_data = _lookup_image(api_client, name)
+    if image_data is None:
+        code, image_data = api_client.images.create_by_url(name, url)
+        assert code == 201, "failed to create image"
+
+        endtime = datetime.now() + timedelta(seconds=wait_timeout)
+        while endtime > datetime.now():
+            code, data = api_client.images.get(name)
+            if 100 == data.get('status', {}).get('progress', 0):
+                break
+            sleep(3)
+        else:
+            raise AssertionError(
+                "Failed to create Image with error:\n"
+                f"Status({code}): {data}"
+            )
+
+    return image_data
+
+
+def _get_ip_from_vmi(vmi):
+    assert _check_vm_ip_assigned(vmi), "virtual machine does not have ip assigned"
+    return vmi['status']['interfaces'][0]['ipAddress']
+
+
+def _check_vm_is_running(vmi):
+    return ('status' in vmi and
+            'phase' in vmi['status'] and
+            'Running' == vmi['status']['phase'] and
+            'nodeName' in vmi['status'] and
+            vmi['status']['nodeName'] != "")
+
+
+def _check_vm_ip_assigned(vmi):
+    return ('status' in vmi and
+            'interfaces' in vmi['status'] and
+            len(vmi['status']['interfaces']) > 0 and
+            'ipAddress' in vmi['status']['interfaces'][0] and
+            vmi['status']['interfaces'][0]['ipAddress'] is not None)
+
+
+def _check_assigned_ip_func(api_client, vm_name):
+    def _check_assigned_ip():
+        code, data = api_client.vms.get_status(vm_name)
+        if code != 200:
+            return False
+        return _check_vm_is_running(data) and _check_vm_ip_assigned(data)
+    return _check_assigned_ip
+
+
+def _wait_for_vm_ready(request, api_client, vm_name):
+    timeout = request.config.getoption('--wait-timeout') or 300
+    endtime = datetime.now() + timedelta(seconds=timeout)
+    while endtime > datetime.now():
+        if _check_assigned_ip_func(api_client, vm_name)():
+            break
+        sleep(5)
+    else:
+        raise AssertionError("Time out while waiting for vm to be created")
+
+
+def _wait_for_write_data(ip, timeout):
+    client = SSHClient()
+    # automatically add host since we only care about connectivity
+    client.set_missing_host_key_policy(AutoAddPolicy)
+
+    def _wait_for_connect():
+        try:
+            client.connect(ip, username=DEFAULT_USER, password=DEFAULT_PASSWORD)
+        except Exception as e:
+            print('Unable to connect to %s: %s' % (ip, e))
+            return False
+
+        script = f"echo {uuid.uuid4().hex} > ~/generate_str; sync"
```

Comment:
try to use `/dev/urandom` to generate a random file and create file md5 checksum to fully test filesystem integrity